New primitive to fill a line with as many doc parts as possible

karl opened the pull request:
This PR adds a new `fill` primitive for laying out documents. This will attempt to fill a line with as many parts of the document as will fit before moving to a new line.

The idea for `fill` comes from page 14 of Wadler's paper: http://homepages.inf.ed.ac.uk/wadler/papers/prettier/prettier.pdf

If this works it could be a possible fix for https://github.com/prettier/prettier/issues/963.

I've aimed for as minimal a change as possible. We keep the existing JSX behaviour of respecting existing new lines. This means that the pretty printing is one way. Once we wrap text we won't ever unwrap it. This seemed like a sensible starting point!

Input:

```
// Wrapping text
x = <div>
  Some text that would need to wrap on to a new line in order to display correctly and nicely
</div>

// Wrapping tags
x = <div>
  <first>f</first> <first>f</first> <first>f</first> <first>f</first> <first>f</first> <first>f</first>
</div>

// Wrapping tags
x = <div>
  <first>f</first><first>f</first><first>f</first><first>f</first><first>f</first><first>f</first>
</div>

// Wrapping tags
x = <div>
  <a /><b /><c />
  <first>aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa</first> <first>f</first>
</div>

// Wrapping tags
x = <div>
  <sadashdkjahsdkjhaskjdhaksjdhkashdkashdkasjhdkajshdkashdkashd /> <first>f</first>
</div>
```

Output:

```
// Wrapping text
x = (
  <div>
    Some text that would need to wrap on to a new line in order to display
    correctly and nicely
  </div>
);

// Wrapping tags
x = (
  <div>
    <first>f</first> <first>f</first> <first>f</first> <first>f</first>
    {" "}
    <first>f</first> <first>f</first>
  </div>
);

// Wrapping tags
x = (
  <div>
    <first>f</first><first>f</first><first>f</first><first>f</first>
    <first>f</first><first>f</first>
  </div>
);

// Wrapping tags
x = (
  <div>
    <a /><b /><c />
    <first>aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa</first>
    {" "}
    <first>f</first>
  </div>
);

// Wrapping tags
x = (
  <div>
    <sadashdkjahsdkjhaskjdhaksjdhkashdkashdkasjhdkajshdkashdkashd />
    {" "}
    <first>f</first>
  </div>
);
```

vjeux commented:
Awesome! @rattrayalex could you take a look to see if that's what you had in mind?

karl commented:
I've just pushed my latest prototyping, and written up a list of important things that still need to be worked on.

It's taken me most of the day to get my head around Walder's paper and the Prettier code base, so I haven't had a chance to tackle the many edge cases I can see will occur with JSX and whitespace!

rattrayalex commented:
`wrap`ing my head around this now...

rattrayalex commented on src/doc-printer.js:
```diff
@@ -176,6 +194,51 @@ function printDocToString(doc, width, newLine) {
+            // console.log('only first fits', [first, second]);
+            cmds.push(remainingCmd)
+            if (typeof second !== 'string' || typeof first !== 'string') {
+              cmds.push([ind, MODE_BREAK, "{' '}"]);
```
woah, this kind of JSX-specific stuff is definitely concerning... perhaps it could come out prior to release...

karl replied:
Yep, this felt strange but it was the easiest way to get something sane printing out!

Ideally the `fill` implementation would be reasonably agnostic, but I don't know enough to make that work yet.

rattrayalex commented on tests/jsx-significant-space/__snapshots__/jsfmt.spec.js.snap:
```diff
@@ -1,199 +1,28 @@
-  <span><Icon icon=\\"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa\\" /> <Icon icon=\\"\\" /></span>;
-
-not_broken_end =
+"x =
```
I'm curious to see what the output is on the existing tests... you're welcome to add a new test fixture while you're developing to avoid running on multiple fixtures.

I know the work so far is pretty preliminary, but any chance you'd be willing to run w/ changes to existing tests too?

karl replied:
I've push up a new commit with all the original tests cases (and my single new test case), including all the change snapshots.

Lots has changed! It looks like most of the changes are due to collapsing newlines in JSX, but there are plenty of bugs as well 😛

rattrayalex commented on src/printer.js:
```diff
@@ -2762,62 +2763,83 @@ function printJSXChildren(path, options, print, jsxWhitespace) {
+        //   words.push(softline);
+        // }
+
+        value.replace(/^\s+|\s+$/g, "").split(/(\s+)/).forEach(word => {
```
I haven't looked at this super in-depth, but it looks like the intent here would be to squash jsx+element mixtures to be as tight as possible?

So, 

```jsx
<div>
  hello
  world
</div>
```
would become 

```jsx
<div>
  hello world
</div>
```
and 

```jsx
<div>
  hello
  <world />
</div>
```
would be come

```jsx
<div>
  hello<world />
</div>
```

?

(maybe you haven't thought this far ahead yet, which is fine – or maybe you plan to go back to the commented-out code soon enough)

karl replied:
I don't really have an intent at the moment, just playing around with what is possible 😀

Currently it is preserving new lines **within** JSX text, but I expect it may compact the join between text and tags.

I was hoping after a period of playing around we'd know enough about the possibilities to come up with a coherent set of layout rules that could then be implemented.

rattrayalex commented on src/doc-printer.js:
```diff
@@ -39,9 +46,20 @@ function fits(next, restCommands, width) {
           break;
+        case "fill":
+          // include number of spaces
+          cmds.push([ind, mode, " ".repeat(doc.parts.length - 1)]);
```
This definitely confuses me, but I'm not super used to the printer.

rattrayalex commented on src/doc-printer.js:
```diff
@@ -176,6 +194,51 @@ function printDocToString(doc, width, newLine) {
+          const firstCmd = [ind, MODE_FLAT, first];
+          const remainingCmd = [ind, MODE_BREAK, fill(remaining)];
+
+          const firstAndSecondCmd = [ind, MODE_FLAT, concat([first, second])];
```
I assume the `firstAndSecond` gambit is primarily intended to understand how to split between text and elements?

karl replied:
The basic idea is copied from fill implementation in Walder's paper (at the end of page 14).

I believe the idea is to check that the first two elements with both fit on a single line before joining them with a space.

If they don't fit on a single line then we need to join them with a new line.

rattrayalex replied:
ah, gotcha, interesting... thanks!

karl commented:
Thanks for taking a look @rattrayalex, very brave jumping in at this early stage 🙇🏻

I'm likely to be working on this in a very on/off fashion over the next few weeks, so don't feel in any hurry with giving feedback. High level approaches/direction stuff if definitely the most useful while I wrap my head around the algorithms/code.

Also, I'm totally happy if this doesn't end up going anywhere. I'm already learning a lot about JSX and pretty printing 😀

rattrayalex commented:
cool, good luck & happy hacking! feel free to reach out, I'd be happy to help!

jlongster commented:
> is flattening something we hope to bring to strings? It'll obviously involve AST-modification, but could be very nice.

It sounds neat, but my argument against it is that we never break strings by splitting one string into multiple (and converting into multiple strings with `+`). We should try to be as symmetrical as possible. Because otherwise it'll collapse something, then you'll add something to it, and it doesn't break it again. The biggest place where this happens now is objects, since once it breaks, it'll never be collapsed again since we keep it expanded. This has caused several stabilization issues since running prettier twice will see the code differently (the first run there were no hard lines, second run there was).

jlongster commented:
This PR looks really neat, thanks for the work! I don't have time to take a close look right now, but this would be great for JSX.

karl commented:
I've managed to get this PR into a reviewable state! 😄

After much time spent wrapping my head around the many different ways of handling whitespace in JSX (so much respect for @rattrayalex having already tamed this!) I finally have a small and reasonably sane approach to filling a line with JSX children.

The new `fill` primitive is now fully generic, so it should be easy to use it for other parts of the document (I think I recall @vjeux wanted to use it for binary expressions?).

I would be great if someone with a little more experience with the code could look at the changes I've made to `doc-printer.js`. Specifically importing and calling the doc builders. It works nicely, but a sanity check would be great!

rattrayalex commented on tests/jsx-significant-space/__snapshots__/jsfmt.spec.js.snap:
```diff
@@ -90,8 +90,9 @@ before = (
-    {" "}
+    <span
+      barbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbarbar
+    />{" "}
```
Hmm, this particular case looks like a regression to me... may not be a big deal.

rattrayalex commented on tests/jsx-significant-space/__snapshots__/jsfmt.spec.js.snap:
```diff
@@ -100,16 +101,14 @@ before_break2 = (
   <span>
-    foo
-    {" "}
+    foo{" "}
```
I think the existing behavior was intentionally added by @vjeux (IIRC, it used to behave as you have it for jsx whitespace)

rattrayalex commented on tests/jsx-significant-space/__snapshots__/jsfmt.spec.js.snap:
```diff
@@ -146,13 +145,15 @@ nest_plz = (
 regression_not_transformed_1 = (
   <span>
-    {" "}<Icon icon="aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa" />
+    {" "}
```
I find this surprising... should mirror the reverse cases, no?

rattrayalex replied:
(eg; similar_2 below)

rattrayalex commented on tests/jsx-significant-space/__snapshots__/jsfmt.spec.js.snap:
```diff
@@ -178,22 +179,15 @@ similar_3 = (
-    {" "}
-    long text long text
+    long text long text long text long text long text long text long text long
+    text <link>url</link> long text long text
```
🎉 🎉 🎉

rattrayalex commented on tests/jsx-split-attrs/__snapshots__/jsfmt.spec.js.snap:
```diff
@@ -179,6 +178,7 @@ leave_opening = (
     submitLabel="Sign in with Google"
   >
     {" "}
+
```
regression

rattrayalex commented on src/doc-printer.js:
```diff
@@ -210,6 +226,77 @@ function printDocToString(doc, options) {
+            break;
+          }
+
+          const split = parts[1];
```
I find the name `split` confusing – is it assumed to always be a whitespace type (`line`, `softline`, etc)?

rattrayalex replied:
might be nice to have a mirroring `splitCmd` variable here, I found `firstCmd` somewhat helpful

rattrayalex replied:
ahh, I think I see... code like this might really clarify things for me as a reader...

```js
const split = parts[1];
assert.ok(split.type === "line" || split.type === "...");
const splitFlatCmd = [ind, MODE_FLAT, split];
const splitBreakCmd = [ind, MODE_BREAK, split];
```

I think technically that's a bit less performant in some cases because not all scenarios require the instantiation of both the Flat and Break cmd's, but the overhead should be quite low. 

I'm also not sure this will properly account for types like `break-parent`, `line-suffix`, `line-suffix-boundary`, etc – which may not appear in the JSX context but are widespread elsewhere. Perhaps that's not worth building for at this point.

rattrayalex commented on src/doc-printer.js:
```diff
@@ -210,6 +226,77 @@ function printDocToString(doc, options) {
+            break;
+          }
+
+          const second = parts[2];
```
`second` seems confusing... it seems to be the third part?

rattrayalex commented on src/doc-printer.js:
```diff
@@ -210,6 +226,77 @@ function printDocToString(doc, options) {
+          }
+
+          const second = parts[2];
+          const remaining = parts.slice(2);
```
do you mean `parts.slice(3)`?

karl commented:
Thanks for the great feedback @rattrayalex!

It might be a week or so before I get to work on this again, but when I do I'll aim at fixing up the regressions and edge cases.

Your right that the fill algorithm could do with a bit of reworking, I was so happy I got it working I submitted the PR without refactoring it 😀

rattrayalex commented:
Awesome! I mean I think what you have is really great, and already very readable (all things considered) ;-)

karl commented:
@rattrayalex I've pushed a hacked together quick fix that (I think) fixes all the regressions 😄

The code could probably do with a tidy up before merging, but I think the behaviour is looking good!

rattrayalex commented on src/printer.js:
```diff
@@ -3024,8 +3029,21 @@ function printJSXElement(path, options, print) {
   const jsxWhitespace = options.singleQuote
-    ? ifBreak("{' '}", " ")
-    : ifBreak('{" "}', " ");
+    ? ifBreak(concat([softline, "{' '}", softline]), " ")
```
the mutations involved later on with this seem pretty bad... but I think you may be on the right track in general. How about something like this: 

1) change this to: `const jsxWhitespace = options.singleQuote ? "{' '}" : '{" "}';`
2) where `jsxWhitespace` is used above, change its usage to `ifBreak(concat([jsxWhitespace, softline]), " ")` (with the appropriate usage of `softline`, typically as indicated in the former code). 

Thoughts?

rattrayalex commented on tests/jsx-split-attrs/__snapshots__/jsfmt.spec.js.snap:
```diff
@@ -122,8 +122,7 @@ long_open_long_children = (
         submitLabel="Sign in with Google"
-      />
-      d
+      />d
```
This really seems wrong to me... is there any way to not use this algorithm if one of the elements has broken?

rattrayalex replied:
If I write `<div with="long attributes..." />blah`, the "blah" should really be on its own line if the div breaks.

karl replied:
Yep, I agree! 

In my first prototype it did have the behaviour of ensuring that the you always had a line break before and after elements that were broken. I must have lost that behaviour in one of my many reworkings of the code.

I'll aim to fix this up, hopefully some time next week.

karl commented:
I've pushed a few more commits that start to tidy up the code around the `fill` algorithm and the JSX whitespace declarations.

I've also fixed a couple of other bugs I found while adding new test cases.

rattrayalex commented on src/printer.js:
```diff
@@ -3127,12 +3143,16 @@ function printJSXElement(path, options, print) {
+    if (child === innerJsxWhitespace) {
       if (i === 0) {
-        groups.unshift(child);
+        if (children.length === 1) {
```
I think it might be a bit more clear if this were pulled out... 

```js
if (child === innerJsxWhitespace) {
  if (children.length === 1) {
    groups.unshift(solitaryJsxWhitespace);
    groups.pop();  // remove unnecessary empty group
    return;
  } else if (i === 0) {
    groups.unshift(leadingJsxWhitespace);
    return;
  } else if (i === children.length - 1) {
    groups.push(trailingJsxWhitespace);
    return;
 }
```

karl replied:
I've now made this tweak!

rattrayalex commented on src/printer.js:
```diff
@@ -3154,6 +3174,16 @@ function printJSXElement(path, options, print) {
   });
 
+  if (children[0] === innerJsxWhitespace) {
+    children[0] = leadingJsxWhitespace;
```
I really don't like these mutations... why does it have to be done this way?

karl replied:
No reason it needs to be done this way, it was just the quickest way for me to test whether the idea of having different leading/trailing whitespace would work.

Very happy to try some different approaches to this. I wonder if the simplest would be to flatten the `groups` array. That way we wouldn't need to duplicate the logic of converting inner whitespace to leading/trailing whitespace.

karl replied:
After thinking about this, I believe this whole chunk of code can be removed.

It seems that `children` is only used in the case where the JSX element is printed on a single line, in which case all the types of JSX whitespaces behave the same, so we don't need to special case leading/trailing/solitary whitespace.

I haven't been able to come up with a failing test case after removing this code, so it seems like a safe change to make.

jlongster commented:
@rattrayalex I haven't looked at this PR for a while, are you implying that we would rewrite comments and strings? I'm very hesitant to do that. Those are things people can format as they wish, and I think we should intentionally avoid touching them and allow people to do what they want. I feel like it would be too prescriptive/aggressive and too many edge cases to try to rewrite those.

rattrayalex commented:
@jlongster thanks for chiming in, I've been curious for your thoughts on that. 

Yes, the thinking would be to rewrite long strings or comments to break across lines. Personally I always find it annoying to do manually and would be very grateful for a tool that would automate the process.

Comments are simpler and lower-risk so I was imagining they'd be an easier place to start. Don't change the AST.

If you haven't heard users clamoring for those use-cases then I definitely don't think it's worth the risk, though. I personally haven't looked through the issues to gauge support recently. 

What are your thoughts on supporting this primitive if it'll only be used for JSX? If you're fine with that, then I think the PR is in a pretty good shape and probably ready for final review, ideally by yourself and/or @vjeux .

karl commented:
While I can't say as to whether we *should* break long strings or comments across lines, the `fill` primitive means we *could* do it.

The `fill` primitive is generic, it just expects an array of alternating **content** and **separator** items. By carefully crafting the separator items it should be possible to split long strings, comments, binary expressions, etc.

### Comments:

The separator would add `// ` to the start of any new lines when splitting a comment.

```
const sep = ifBreak(concat([line, '// ']), ' ');
fill(['// first', sep, 'second', sep, 'third']);
```

### Strings:

The separator would add `" ` to the end of the current line and `+ "` to the start of the next line when splitting strings.

```
const sep = ifBreak(concat(['"', line, '+ "']), ' ');
fill(['"first', sep, 'second', sep, 'third"']);
```

### Binary expressions:

The easy case! We just use `line` as the separator as it will use a space when items fits on one line and a line break when they won't

```
const sep = line;
fill(['a', sep, '+', sep, 'b']);
```

So we'd have the ability to split long comments and strings, we'd just need to decide if we wanted to! 

One thing to note is that `fill` only takes care of splitting, there is currently no mechanism for recombining if we find a split is no longer required.

karl commented:
Hi @jlongster / @vjeux I believe this PR is awaiting final review by either of you. 

I'm in no rush to get it merged (it will be another week at least before we are ready to adopt Prettier at work. I just wondered if there is anything else I can do to make your lives easier when reviewing it?

vjeux commented:
Soooo sorry for the super late review, I've been trying to do all the things needed for Facebook adoption and didn't get the chance to look at this one. Doing it now.

vjeux commented on src/printer.js:
```diff
@@ -3375,16 +3387,18 @@ function printJSXChildren(path, options, print, jsxWhitespace) {
         // eg; one or more spaces separating two elements
         for (let i = 0; i < value.length; ++i) {
-          children.push(jsxWhitespace);
+          children.push(innerJsxWhitespace);
```
Oops, I attempted to fix the merge conflict but it didn't work, it's now breaking many tests

vjeux commented:
Hey, I've been trying to run it on fb codebase but I'm not sure i'm getting the right results because of the merge conflict. In the way I've solved it, the tests are no longer passing and I get some bad results like:

![image](https://cloud.githubusercontent.com/assets/197597/25764014/ba9d1da8-319a-11e7-8606-3d78ca7c40c5.png)

So it doesn't rebase cleanly :(

Would you mind doing a clean rebase such that I can actually review it? So sorry!

karl commented:
No worries!

I should have some time to rebase it early next week and then I'll ping you for a review.

vjeux commented:
By the way, I wanted to mention that I'm very supportive of this pull request. I just want to run it across fb codebase and make sure it doesn't trigger bad side effects.

karl commented:
@vjeux It was quicker than I thought to rebase, so this PR should be ready for review again 😀

vjeux commented:
Looking, thanks!

vjeux commented:
Would it be possible to transform

```js
const Abc = () => {
  return (
    <div>
      Please state your
      {" "}
      <b>name</b>
      {" "}
      and
      {" "}
      <b>occupation</b>
      {" "}
      for the board of directors.
    </div>
  );
};
```

into

```js
const Abc = () => {
  return (
    <div>
      Please state your <b>name</b> and <b>occupation</b> for the board of directors.
    </div>
  );
}
```

?

vjeux commented:
```js
<fbt>
  This has the best cost per
  <fbt:param name="this is a very long string that happens to overflow 80 columns.">
    {label}
  </fbt:param>
</fbt>
```

outputs

```js
<fbt>
  This
  has
  the
  best
  cost
  per
  <fbt:param name="this is a very long string that happens to overflow 80 columns.">
    {label}
  </fbt:param>
</fbt>;
```

which looks very weird, and happens quite a lot inside of fb codebase.

vjeux commented:
An empty line is being inserted between words

```js
<fbt>
  <fbt:param name="cost">
    {this.getCostPerValue(value, currency)}
  </fbt:param>
  per
  <fbt:param name="objective">
    {costPerLabel}
  </fbt:param>
 </fbt>
```

turns into

```js
<fbt>
  <fbt:param name="cost">
    {this.getCostPerValue(value, currency)}
  </fbt:param>

  per

  <fbt:param name="objective">
    {costPerLabel}
  </fbt:param>
</fbt>;
```

vjeux commented:
```js
<fbt>
  <fbt:param name="1st friend">
    {names[0]}
  </fbt:param>,
  <fbt:param name="2nd friend">
    {names[1]}
  </fbt:param>
  and
  <fbt:param name="number of friends who have been here minus 2">
    {count - 2}
  </fbt:param>
  other friends have been here
</fbt>
```

outputs

```js
<fbt>
  <fbt:param name="1st friend">
    {names[0]}
  </fbt:param>
  ,

  <fbt:param name="2nd friend">
    {names[1]}
  </fbt:param>

  and

  <fbt:param name="number of friends who have been here minus 2">
    {count - 2}
  </fbt:param>
  other friends have been here
</fbt>;
```

Notice how the `,` is no longer on the same line.

vjeux commented:
I'm very impressed! Outside of the 2 issues in the middle (which trigger a lot), it looks really awesome! It inlines a bunch of long descriptions :)

If you can make #1 work that would be great, so all the code that has been converted by prettier in the past can make use of those improvements.

For #4, I could go either way, it would be best if it was on the same line but I can live with it being on the next line.

Let me know how you want to proceed to fix the middle two. Once they are fixed, I'll ship it :)

karl commented:
Thanks for the extra test cases! 

I totally agree that the middle two are bugs and I'll get those fixed up.

Number 1 is a major change to how we handle JSX, that I know @rattrayalex had some reservations around changing. Currently we preserve all new lines in JSX, and automatically collapsing lines would mean we'd need to re-think that.

I'm really interested in tackling this, but I think another PR would be best as I suspect it will engender a good bit of discussion!

I imagine we could special case the trailing comma with a rule such as allow 1 character text to follow a group with a break. I'd like to tackle that in a separate PR as well.

azz commented:
I actually like the `,` on a separate line.

Awesome work with this PR, btw 🏅

karl commented:
@vjeux I've fixed up the middle two bugs (words incorrectly output one per line, and extra blank lines.

In the process of investigating how we decide to render the `,` on a new line I found and fixed another issue where we would occasionally output text on the same line as a element that rendered across multiple lines. Now we always have a new line after a child element that is rendered across multiple lines. It turns out doing this simplified a chunk out the JSX rendering!

grydstedt commented:
@karl Testing out your branch to see if I can get rid of some {' '}. Great work!

I noticed one thing, not sure if it's intended:

![screen shot 2017-05-09 at 2 37 43 pm](https://cloud.githubusercontent.com/assets/782394/25873937/2db56908-34c5-11e7-97fd-049dbff70a4e.png)

becomes

![screen shot 2017-05-09 at 2 39 14 pm](https://cloud.githubusercontent.com/assets/782394/25873961/4bb5162e-34c5-11e7-9432-165df746243b.png)

(it adds the {' '}). However, this does not happen if aVariable is replaced by text.

vjeux commented:
It sounds like `{' '}tag in it` can be on the same line

karl commented:
Having the`{' '}` always be on a line by itself was the easiest way to implement the JSX white space. It works well when separating content that ends up being multi line. I wonder if we could tweak the layout in certain situations.

Do you think this is something that would block the PR?

vjeux commented:
No, I'm happy to merge with this. Sorry I'm oncall this week and need to find time to run it through again.

vjeux commented:
So so so sorry for the super long wait, I just ran this on our codebase and almost nothing changed, so I'm happy. Thanks for fixing all the edge cases :)

vjeux commented:
For the issue with the comma, what about the following heuristic: if there is no space between the text and the jsx tag, then keep it that way?

This is also an issue for `(<tag />)`

```diff
-        (<fbt:param name="Number of profile items to update">
+        (
+        <fbt:param name="Number of profile items to update">
           {itemCount}
-        </fbt:param>)
+        </fbt:param>
+        )
```

karl commented:
Thanks for finding the time to merge this @vjeux! And thanks to you and @rattrayalex for patiently taking me through my first major PR, it's been a pleasure 😀

karl commented:
☝🏻 I've opened a few issue to discuss possible follow on work from this PR.

pkarl commented:
Awesome!
